### PR TITLE
Add `extensions` option for TypeScript support and other extensions in the future

### PR DESCRIPTION
**Pull request by huan**
Follow https://github.com/avajs/ava/issues/1109

### CODE REVIEW

**Comment by sindresorhus** [comment]
There are more places here where `.js` is hardcoded. Also see my comment here: https://github.com/avajs/ava-files/pull/4#issuecomment-246761791

We also need to decide whether `extension` should replace `.js` or be an addition.

**Comment by zixia** [comment]
Ah, got it.

I just did a check, and list the following parts need to add `extensions` support too:

* [ ] `defaultIncludePatterns()`
* [ ] `mixedPatterns`
* [ ] `isTest()`
    * [ ] `path.extname(filePath)`
    * [ ] `recursivePatterns`
* [ ] `getChokidarPatterns()`
* [ ] `handlePaths()`

About whether `extensions` should replace `.js` or be an addition, my idea is **replace**. That's because I believe `extensions` should include `.js` in it, and without hardcoded `.js`, the code can be cleaner for support extensions.

Which one do you prefer, *replace* or *additional*?

**Comment by sindresorhus** [comment]
Thinking about it more, replacing makes the most sense here. Why would you even use `.js` if you have a TypeScript project, and if you have that edge-case, it's still possible to manually specify it.

**Comment by zixia** [comment]
Ok, so both of us agree to use **replace**. 

I'll try to modify the code later, get rid of all part of the *hard coded .js*, and start using `extensions` to rewrite current related codes.

File: index.js [hunk 1]
```diff
@@ -241,14 +266,22 @@ function handlePaths(files, excludePatterns, globOptions) {
...
+				var patternList = [];
+				globOptions.extensions.forEach(function (ext) {
+					patternList.push(path.join(file, '**', '*.' + ext));
+				});
 
 				if (process.platform === 'win32') {
 					// Always use / in patterns, harmonizing matching across platforms.
-					pattern = slash(pattern);
+					// pattern = slash(pattern);
+					patternList = patternList.map(function (p) {
```
Comment (jedmao) [inline index.js]: This should be the same as `patternList.map(slash);` no?

Comment (zixia) [inline index.js]: yes, you are right, it's the same. thanks!

File: index.js [hunk 2]
```diff
@@ -17,14 +17,21 @@ function defaultExcludePatterns() {
 	];
 }
 
-function defaultIncludePatterns() {
-	return [
-		'test.js',
-		'test-*.js',
+function defaultIncludePatterns(extensions) {
+	var patternList = [
```
Comment (vdemedes) [inline index.js]: Imo it could just be `patterns`.

File: index.js [hunk 3]
```diff
@@ -241,14 +266,20 @@ function handlePaths(files, excludePatterns, globOptions) {
 
 				searchedParents[file] = true;
 
-				var pattern = path.join(file, '**', '*.js');
+				// var pattern = path.join(file, '**', '*.js');
+				var patternList = [];
```
Comment (vdemedes) [inline index.js]: `patterns` too?

File: index.js [hunk 4]
```diff
@@ -241,14 +266,20 @@ function handlePaths(files, excludePatterns, globOptions) {
...
+				globOptions.extensions.forEach(function (ext) {
+					patternList.push(path.join(file, '**', '*.' + ext));
```
Comment (vdemedes) [inline index.js]: Inconsistency between https://github.com/avajs/ava-files/pull/9/files#diff-168726dbe96b3ce427e7fedce31bb0bcR187. `path.join()` is used here, unlike in that other place. I think it should be used in both places.

Comment (vdemedes) [inline index.js]: Although, since we are using `slash()`, `path.join()` is not necessary.

File: index.js [hunk 5]
```diff
@@ -269,6 +300,16 @@ function handlePaths(files, excludePatterns, globOptions) {
 		});
 }
 
+function validFileName(file, extensions) {
```
Comment (vdemedes) [inline index.js]: `isValidFileName` would be better.

File: test.js [hunk 6]
```diff
@@ -139,3 +139,21 @@ test('findFiles - finds the correct files by default', async t => {
...
+	const fixtureDir = fixture('ext');
+	const avaFiles = new AvaFiles({
+		files: ['**/*.ts'],
+		cwd: fixtureDir,
+		extensions: ['ts']
+	});
+
+	const files = await avaFiles.findTestFiles();
+	const expected = [
+		'typescript.ts'
```
Comment (vdemedes) [inline test.js]: Can be on the same line.

File: test.js [hunk 7]
```diff
@@ -139,3 +139,21 @@ test('findFiles - finds the correct files by default', async t => {
...
+	].map(function (file) {
```
Comment (vdemedes) [inline test.js]: Should be an arrow function.

File: index.js (hunk 2 above)
Comment (zixia) [inline index.js]: done.

File: index.js (hunk 3 above)
Comment (zixia) [inline index.js]: done too.

File: index.js (hunk 4 above)
Comment (zixia) [inline index.js]: done

File: index.js (hunk 5 above)
Comment (zixia) [inline index.js]: done

File: test.js (hunk 6 above)
Comment (zixia) [inline test.js]: done

File: test.js (hunk 7 above)
Comment (zixia) [inline test.js]: done

**Comment by zixia** [comment]
Hey @vdemedes ,

Thanks for the review, I had followed your suggestions and made a new push.

**Comment by sindresorhus** [comment]
I would be more confident in this change if it had more tests.

**Comment by sindresorhus** [comment]
There's lot of commented code left in there. Please always review the diff before pushing changes.

**Comment by zixia** [comment]
Sorry for the commented code because I use them to compare the code at first, and forgot to clean them. Just cleaned them all.

About the tests, I added a new test with empty `extensions` option, which should get nothing from ava-files. 

What kind of tests do you think we need to add more? I can follow you, but now I can only think about test `extensions` parameter, which could be a `['ts']` or `[]`.

**Comment by novemberborn** [comment]
@zixia this might prove quite useful for [RFC001](https://github.com/avajs/ava/blob/master/docs/specs/001%20-%20Improving%20language%20support.md)! However we've decided to move this code back into AVA core. I'm closing this PR since it won't land here, but I'll make sure to remember it for when we need this functionality.

**Pull request closed without merging**